pineappl_capi/include/PineAPPL.hpp

```diff
@@ -119,15 +115,79 @@ struct Lumi {
   }
 
   /**
-   * @brief Returns the number of combinations of the luminosity function `lumi`
-   * for the specified entry.
+   * @brief Returns the number of combinations of the luminosity function
+   * `lumi` for the specified entry.
    * @param entry position in lumi
    */
   std::size_t combinations(const std::size_t entry) const {
     return pineappl_lumi_combinations(this->raw, entry);
   }
 };
 
+/** @brief Entry in sub-channel function.  */
+struct SubChannelEntry {
+  /** @brief First parton id. */
+  std::vector<std::pair<std::vector<int>, double>> entry;
+};
+
+/** @brief Entry in channels function.  */
+struct ChannelsEntry {
+  /** @brief First parton id. */
+  std::vector<SubChannelEntry> channels_entry;
+};
+
+/** @brief Channels function. */
+struct Channels {
+  /** @brief Underlying raw object. */
+  pineappl_channels *raw;
+
+  /** @brief Constructor. */
+  Channels() : raw(pineappl_channels_new()) {}
+  Channels(const Channels &) = delete;
+  Channels(Channels &&) = delete;
+
+  Channels &operator=(const Channels &) = delete;
+  Channels &operator=(Channels &&) = delete;
+
+  /** @brief Destructor. */
+  ~Channels() { pineappl_channels_delete(this->raw); }
+
+  /** @brief Number of elements. */
+  std::size_t count() const { return pineappl_channels_count(this->raw); }
+
+  /**
+   * @brief Add a channel function.
+   * @param c channel function
+   */
+  void add(const ChannelsEntry &c) const {
+    const std::size_t combinations = c.channels_entry.size();
+    const std::size_t nb_convolutions =
```

Comment:
```suggestion
    if (0 == combinations)
      return;
    const std::size_t nb_convolutions =
```
and we need more similar things - since in C++ people usually do sanity checks (unlike Python). E.g. make sure `c.channels_entry[0].entry[0]` exists, and all entries have the same size (can e.g. be done during the loop below)

Reply:
Yes, that'd be the plan indeed. And these are the easiest parts of the whole process 😇